hydra_base/lib/HydraTypes/Types.py

```diff
@@ -0,0 +1,228 @@
+
+    def set_value(self, val):
+        self._value = val
+
+    value = property(get_value, set_value)
+
+
+class Array(DataType):
+    tag      = "ARRAY"
+    skeleton = "[%f, ...]"
+    json     = ArrayJSON()
+
+    def __init__(self, encstr):
+        self.value = encstr
+        self.validate()
+
+    @classmethod
+    def fromDataset(cls, value, metadata=None):
+        return cls(value)
+
+    def validate(self):
+        j = json.loads(self.value)
+        assert len(j) > 0           # Sized
+        assert iter(j) is not None  # Iterable
+        assert j.__getitem__        # Container
+        assert not isinstance(j, six.string_types) # Exclude strs
+
+    def get_value(self):
+        return self._value
+
+    def set_value(self, val):
+        self._value = val
+
+    value = property(get_value, set_value)
+
+
+
+class Descriptor(DataType):
+    """ Unused obsolete type """
+    tag      = "DESCRIPTOR"
+    skeleton = "%s"
+    json     = DescriptorJSON()
+
+    def __init__(self, data):
+        self.value = data
+        self.validate()
+
+    @classmethod
+    def fromDataset(cls, value, metadata=None):
+        if metadata and metadata.get('data_type') == 'hashtable':
+            try:
+                df = pd.read_json(six.text_type(value))
+                data = df.transpose().to_json()
+            except Exception:
+                noindexdata = json.loads(six.text_type(value))
+                indexeddata = {0:noindexdata}
+                data = json.dumps(indexeddata)
+            return cls(data)
+        else:
+            return cls(six.text_type(value))
+
+
+    def validate(self):
+        pass
+
+    def get_value(self):
+        return self._value
+
+    def set_value(self, val):
+        self._value = val
+
+    value = property(get_value, set_value)
+
+
+class Dataframe(DataType):
+    tag      = "DATAFRAME"
+    skeleton = "%s"
+    json     = DataframeJSON()
+
+    def __init__(self, data):
+        self.value = data
+        self.validate()
+
+    @classmethod
+    def fromDataset(cls, value, metadata=None):
+        try:
+            ordered_jo = json.loads(six.text_type(value), object_pairs_hook=collections.OrderedDict)
+            df = pd.DataFrame.from_dict(ordered_jo)
+        except ValueError as e:
+            """ Raised on scalar types used as pd.DataFrame values
+                in absence of index arg
+            """
+            raise HydraError(e.message)
+
+        return cls(df)
+
+
+    def validate(self):
+        assert isinstance(self._value, pd.DataFrame)
+        assert not self._value.empty
+
+
+    def get_value(self):
+        return self._value.to_json()
+
+    def set_value(self, val):
+        self._value = val
+
+    value = property(get_value, set_value)
+
+
+class Timeseries(DataType):
+    tag      = "TIMESERIES"
+    skeleton = "[%s, ...]"
+    json     = TimeseriesJSON()
+
+    def __init__(self, ts):
+        self.value = ts.to_json(date_format='iso', date_unit='ns')
```

Comment:
Can you make this look like the dataframe structure, whereby the to_json is set in the get_value, and the object deals with dataframes for _value and not strings.